### PR TITLE
Elaborate on types in unions with the most overlap in properties

### LINKED ISSUE #26450: Seek most-overlapping types when relating object types in unions
From @rkirov at https://github.com/DefinitelyTyped/DefinitelyTyped/issues/28044#issue-349690446

```ts
interface A {
    a: string;
    b: number
};

interface Other {
    o: number;
}

declare function f(x: A | Other): any;

export let x = { a: '', b: '' };
export let y = f(x);
```

**Expected**

```
Argument of type '{ a: string; b: string; }' is not assignable to parameter of type 'Other | A'.
  Type '{ a: string; b: string; }' is not assignable to type 'A'.
    Types of property 'b' are incompatible.
      Type 'string' is not assignable to type 'number'.
```

**Actual**

```
Argument of type '{ a: string; b: string; }' is not assignable to parameter of type 'A | Other'.
  Type '{ a: string; b: string; }' is not assignable to type 'Other'.
    Property 'o' is missing in type '{ a: string; b: string; }'.
```

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
When relating a type *S* to a union type *T*, if *S* is an object type or intersection type, then we will find the last object or union type *T'* in *T* which shares the most properties with *S*.

Fixes #26450

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -11512,6 +11513,32 @@ namespace ts {
                 }
             }
 
+            function findMostOverlappyType(source: Type, unionTarget: UnionOrIntersectionType) {
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: OverlappyType is a new favorite term.

Comment (DanielRosenwasser) [reply src/compiler/checker.ts]: Another candidate was `getMostShallowSatisfiedType` but that sounded slightly judgemental.

Comment (DanielRosenwasser) [reply src/compiler/checker.ts]: So these types are always sorted by their ID, and those are always based on declaration order.

For named types (e.g. interface declarations) that's not usually that useful, but sometimes there's an intent when the user writes out anonymous types. For example, if the user wrote type literals such as `{ someProp: number } | { anotherProp: string }`, my own experience has shown me that a lot of the time the first one actually *is* thought of as the more common case. That's part of why I went with the first type in #26895 (though I think keeping it simple was the primary motivator).

When I tried this and only took the first most-overlappy type, I think I found a decent number of hit-or-miss changes. I didn't see significant improvements in the baselines and given that the change is literally switching between a `>` and a `>=`, I'd rather keep the baselines cleaner in that point. I think the high order bit is that we just look for overlappy types in the first place.

Comment (DanielRosenwasser) [reply src/compiler/checker.ts]: Right, in our experience it ended up as more of a lavash or a naan, but I wouldn't be surprised if not sorting the IDs made type comparisons a pita.

Comment (CyrusNajmabadi) [reply src/compiler/checker.ts]: Barbari or Sangak 4ever.

Comment (RyanCavanaugh) [reply src/compiler/checker.ts]: Please help @JsonFreeman

Comment (weswigham) [inline src/compiler/checker.ts]: So,I think this is a slightly cleaner implementation that should also handle generics better: 
```ts
function findMostOverlappyType(source: Type, unionTarget: UnionOrIntersectionType) {
    let bestMatch = unionTarget.types[0];
    let matchingCount = 0;
    for (const target of unionTarget.types) {
        const overlap = getIntersectionType([getIndexType(source), getIndexType(target)]);
        if (overlap.kind & TypeFlags.Index) {
            // perfect overlap of keys
            bestMatch = target;
            matchingCount = Infinity;
        }
        if (overlap.flags & TypeFlags.Union) {
            // Some subset overlap
            const len = length((overlap as UnionType).types);
            if (len >= matchingCount) {
                bestMatch = target;
                matchingCount = len;
            }
        }
    }
    return bestMatch;
}
```

Comment (JsonFreeman) [reply src/compiler/checker.ts]: I live in NYC now so I have to vote for bagels

File: tests/baselines/reference/checkJsxChildrenProperty2.errors.txt [hunk 2]
```diff
@@ -3,19 +3,19 @@ tests/cases/conformance/jsx/file.tsx(17,11): error TS2710: 'children' are specif
...
 tests/cases/conformance/jsx/file.tsx(43,6): error TS2322: Type '{ children: (string | Element)[]; a: number; b: string; }' is not assignable to type 'Prop'.
   Types of property 'children' are incompatible.
     Type '(string | Element)[]' is not assignable to type 'string | Element'.
-      Type '(string | Element)[]' is missing the following properties from type 'Element': type, props
+      Type '(string | Element)[]' is not assignable to type 'string'.
 tests/cases/conformance/jsx/file.tsx(49,6): error TS2322: Type '{ children: Element[]; a: number; b: string; }' is not assignable to type 'Prop'.
   Types of property 'children' are incompatible.
     Type 'Element[]' is not assignable to type 'string | Element'.
-      Type 'Element[]' is missing the following properties from type 'Element': type, props
+      Type 'Element[]' is not assignable to type 'string'.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/checkJsxChildrenProperty2.errors.txt]: Regression :(

Comment (weswigham) [reply tests/baselines/reference/checkJsxChildrenProperty2.errors.txt]: It's probably better this way, for now, until the element -> argument case has a special handler.

**Pull request merged**